README.md
<h1 align="center">Firebase Auth</h1>
<p align="center">
 Firebase/Admin Auth Javascript Library for Cloudflare Workers
</p>
<br>

<p align="center">
  <a href="https://github.com/divyam234/firebase-auth-cfworkers/blob/main/LICENSE"><img alt="License" src="https://img.shields.io/github/license/divyam234/firebase-auth-cfworkers"/></a>
  <img src="https://github.com/divyam234/firebase-auth-cfworkers/actions/workflows/node_ci.yaml/badge.svg" alt="GitHub CI"/>
  <a href="https://www.npmjs.com/package/firebase-auth-cfworkers"><img alt="NPM" src="https://badge.fury.io/js/firebase-auth-cfworkers.svg"/></a>
  <a href="https://www.npmjs.com/package/firebase-auth-cfworkers"><img src="https://img.shields.io/npm/dt/firebase-auth-cfworkers.svg" alt="NPM Downloads"/></a>
  <a href="https://github.com/divyam234><img alt="Github" src="https://img.shields.io/static/v1?label=GitHub&message=divyam234&color=005cb2"/></a>
</p>

# Supported operations:

- [x] createSessionCookie()
- [x] verifySessionCookie()
- [x] signInWithEmailAndPassword()
- [x] signUpWithEmailAndPassword()
- [x] changePassword()
- [x] lookupUser()
- [x] setCustomUserClaims()

# Install

```bash
npm i firebase-auth-cfworkers
```

# Usage

Firebase tries to use the same method names and return values as the official Firebase/Admin SDK. Sometimes, the method signature are slightly different.

**Create FirebaseAuth**

```ts
import { FirebaseAuth } from 'firebase-auth-cfworkers';

const auth = new FirebaseAuth({
  apiKey: 'Firebase api key',
  projectId: 'Firebase project id',
  privateKey: 'Firebase private key or service account private key',
  serviceAccountEmail: 'Firebase service account email',
});
```

**Sign-in with email/pass**

```ts
//Sign in with username and password
const { token, user } = await auth.signInWithEmailAndPassword(
  'my@email.com',
  'supersecurepassword'
);

const userEmail = user.email;
const refreshToken = token.refreshToken;
```

**Sign-up with email/pass**

```ts
//Sign up with username and password
const { token, user } = await auth.signUpWithEmailAndPassword(
  'my@email.com',
  'supersecurepassword'
);

const userEmail = user.email;
const refreshToken = token.refreshToken;
```

**Set Custom User Claims**

```ts
//Set Custom User Claims
const res = await auth.setCustomUserClaims(uid,
  {'admin':true}
);
```

**Create session cookies**

```ts
//Create a new session cookie from the user idToken
const { token, user } = await auth.signInWithEmailAndPassword(
  'my@email.com',
  'supersecurepassword'
);

const sessionCookie = await auth.createSessionCookie(token.idToken);
```

**Verify session cookies**

```ts
auth
  .verifySessionCookie(sessionCookie)
  .then((token) => useToken(token))
  .catch((e) => console.log('Invalid session cookie'));
```

**Cache OAuth tokens with CloudflareKv KV**

```ts
import { FirebaseAuth, CloudflareKv } from 'firebase-auth-cfworkers';

const auth = new FirebaseAuth({
  apiKey: 'Firebase api key',
  projectId: 'Firebase project id',
  privateKey: 'Firebase private key or service account private key',
  serviceAccountEmail: 'Firebase service account email',
  cache: new CloudflareKv(NAMESPACE),
});
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.2.3](https://github.com/divyam234/firebase-auth-cfworkers/compare/v1.2.2...v1.2.3) (2022-11-26)

### 1.2.2 (2022-11-26)

### 1.2.1 (2022-11-26)


src/lib/google-oauth.ts
import {
  decodeProtectedHeader,
  importPKCS8,
  importX509,
  JWTPayload,
  jwtVerify,
  SignJWT,
} from 'jose';
import axios from 'redaxios';

/**
 * Get an OAuth 2.0 token from google authentication apis using
 * a service account
 * @param serviceAccountEmail Email of the service account
 * @param privateKey Private key of the service account
 * @param scope scope to request
 * @returns OAuth 2.0 token
 */
export async function getAuthToken(
  serviceAccountEmail: string,
  privateKey: string,
  scope: string
): Promise<string> {
  const ecPrivateKey = await importPKCS8(privateKey, 'RS256');

  const jwt = await new SignJWT({ scope: scope })
    .setProtectedHeader({ alg: 'RS256' })
    .setIssuer(serviceAccountEmail)
    .setAudience('https://oauth2.googleapis.com/token')
    .setExpirationTime('1h')
    .setIssuedAt()
    .sign(ecPrivateKey);

  const headers = { 'Content-Type': 'application/x-www-form-urlencoded' };

  const payload = new URLSearchParams({
    grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
    assertion: jwt,
  });

  const res = await axios.post('https://oauth2.googleapis.com/token', payload, {
    headers,
  });
  return res.data.access_token;
}

/**
 * Verifies an Identity Platform ID token.
 * If the token is valid, the promise is fulfilled with the token's decoded claims; otherwise, the promise is rejected.
 * @param idToken An Identity Platform ID token
 * @param customData Public Key Data
 */
export async function verifyIdToken(
  idToken: string,
  customData: Record<string, unknown> = {}
): Promise<JWTPayload> {
  //Fetch public keys
  //TODO: Public keys should be cached until they expire
  let data = {};
  if (Object.keys(customData).length > 0) data = customData;
  else {
    const res = await axios.get(
      'https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com'
    );
    data = res.data;
  }

  //Get the correct publicKey from the key id
  const header = decodeProtectedHeader(idToken);
  const certificate = data[header.kid];
  const publicKey = await importX509(certificate, 'RS256');

  //Verify JWT with public key
  const { payload } = await jwtVerify(idToken, publicKey);
  return payload;
}


src/lib/flarebase-auth.ts
import { decodeProtectedHeader, importX509, jwtVerify } from 'jose';
import axios from 'redaxios';

import { Cache } from './cache/cache';
import { getAuthToken, verifyIdToken } from './google-oauth';
import { DecodedIdToken, User } from './models';

export type FlarebaseConfig = {
  readonly projectId: string;
  readonly apiKey: string;
  readonly serviceAccountEmail: string;
  readonly privateKey: string;
  readonly cache?: Cache;
};

/**
 * Interact with Firebase REST Api and Google Identity Toolkit Api.
 * Made to work with Cloudflare Workers
 */
export class FlarebaseAuth {
  private BASE_URL = 'https://identitytoolkit.googleapis.com/v1/';

  constructor(public readonly config: FlarebaseConfig) {}

  /**
   * Cache the result of an async function
   * @param action Function with result to be stored
   * @param key Where to find/store the value from/to the cache
   * @param expiration Cache expiration in seconds
   * @returns Cached result
   */
  private async withCache<T>(
    action: () => Promise<T>,
    key: string,
    expiration: number
  ): Promise<T> {
    if (!this.config.cache) return await action();

    let result = (await this.config.cache.get(key)) as T;
    if (!result) {
      result = await action();
      await this.config.cache.put(key, result, { expirationTtl: expiration });
    }

    return result;
  }

  /**
   * Cache Result Of Token Response
   * @returns Returns Cached result
   */

  async getToken(): Promise<string> {
    return await this.withCache(
      () =>
        getAuthToken(
          this.config.serviceAccountEmail,
          this.config.privateKey,
          'https://www.googleapis.com/auth/identitytoolkit'
        ),
      'google-oauth',
      3600
    );
  }

  /**
   * Send a post request to the identity toolkit api
   * @param formData POST form data
   * @param endpoint endpoint of the identity toolkit googleapis
   * @returns HTTP Response
   */
  private async sendFirebaseAuthPostRequest(
    formData: Record<string, string>,
    endpoint: string
  ): Promise<Record<string, unknown>> {
    const URI =
      this.BASE_URL + `accounts:${endpoint}?key=${this.config.apiKey}`;

    const res = await axios.post(URI, formData);
    if (res.status != 200) throw Error(res.data);
    return res.data;
  }

  /**
   * Retrieve user info from a Firebase ID token
   * @param idToken A valid Firebase ID token
   * @returns User info linked to this ID token
   */
  public async lookupUser(idToken: string): Promise<User> {
    const data = await this.sendFirebaseAuthPostRequest(
      { idToken: idToken },
      'lookup'
    );

    return data.users[0] as User;
  }

  /**
   * Sign in Firebase user with email and password
   * @param email Email of the Firebase user
   * @param password Password of the Firebase user
   * @returns The decoded JWT token payload and the signed in user info
   */
  async signInWithEmailAndPassword(
    email: string,
    password: string
  ): Promise<{ token: DecodedIdToken; user: User }> {
    const data = await this.sendFirebaseAuthPostRequest(
      {
        email: email,
        password: password,
        returnSecureToken: 'true',
      },
      'signInWithPassword'
    );

    const token = data as DecodedIdToken;
    const user = await this.lookupUser(token.idToken);

    return { token, user };
  }

  /**
   * Change a user's password
   * @param idToken	A Firebase Auth ID token for the user.
   * @param newPassword	User's new password.
   * @returns The decoded JWT token payload
   */
  async changePassword(
    idToken: string,
    newPassword: string
  ): Promise<DecodedIdToken> {
    const data = await this.sendFirebaseAuthPostRequest(
      {
        idToken: idToken,
        password: newPassword,
        returnSecureToken: 'true',
      },
      'update'
    );

    const token = data as DecodedIdToken;

    return token;
  }

  /**
   * Delete a current user
   * @param idToken	A Firebase Auth ID token for the user.
   */
  async deleteAccount(idToken: string) {
    await this.sendFirebaseAuthPostRequest(
      {
        idToken: idToken,
      },
      'delete'
    );
  }

  /**
   * Sign up Firebase user with email and password
   * @param email Email of the Firebase user
   * @param password Password of the Firebase user
   * @returns The decoded JWT token payload and the signed in user info
   */
  async signUpWithEmailAndPassword(
    email: string,
    password: string
  ): Promise<{ token: DecodedIdToken; user: User }> {
    const data = await this.sendFirebaseAuthPostRequest(
      {
        email: email,
        password: password,
        returnSecureToken: 'true',
      },
      'signUp'
    );

    const token = data as DecodedIdToken;
    const user = await this.lookupUser(token.idToken);

    return { token, user };
  }

  /**
   * Sets additional developer claims on an existing user identified by the
   * provided `uid`, typically used to define user roles and levels of
   * access. These claims should propagate to all devices where the user is
   * already signed in (after token expiration or when token refresh is forced)
   * and the next time the user signs in.
   *
   * See {@link https://firebase.google.com/docs/auth/admin/custom-claims |
   * Defining user roles and access levels}
   * for code samples and detailed documentation.
   *
   * @param uid - The `uid` of the user to edit.
   * @param customUserClaims - The developer claims to set. If null is
   *   passed, existing custom claims are deleted. Passing a custom claims payload
   *   larger than 1000 bytes will throw an error. Custom claims are added to the
   *   user's ID token which is transmitted on every authenticated request.
   *   For profile non-access related user attributes, use database or other
   *   separate storage systems.
   * @returns A promise that resolves when the operation completes
   *   successfully.
   */
  public async setCustomUserClaims(
    uid: string,
    customUserClaims: Record<string, unknown> | null
  ): Promise<Record<string, unknown>> {
    if (customUserClaims === null) {
      customUserClaims = {};
    }

    const token = await this.getToken();

    const payload = {
      localId: uid,
      customAttributes: JSON.stringify(customUserClaims),
    };

    const headers = { Authorization: 'Bearer ' + token };

    const path = `projects/${this.config.projectId}/accounts:update`;

    const res = await axios.post(this.BASE_URL + path, payload, { headers });
    if (res.status != 200) throw Error(res.data);

    return res.data;
  }

  /**
   * Creates a session cookie for the given Identity Platform ID token.
   * The session cookie is used by the client to preserve the user's login state.
   * @param idToken A valid Identity Platform ID token
   * @param expiresIn The number of seconds until the session cookie expires.
   * Specify a duration in seconds, between five minutes and fourteen days, inclusively.
   * @returns The session cookie that has been created
   */
  async createSessionCookie(
    idToken: string,
    expiresIn: number = 60 * 60 * 24 * 14 //14 days
  ): Promise<string> {
    //Create the OAuth 2.0 token
    //OAuth token is cached until expiration (1h)
    const token = await this.getToken();

    //Post params and header authorization

    const payload = { idToken: idToken, validDuration: expiresIn + '' };
    const headers = { Authorization: 'Bearer ' + token };

    const path = `projects/${this.config.projectId}:createSessionCookie`;
    const res = await axios.post(this.BASE_URL + path, payload, { headers });
    if (res.status != 200) throw Error(res.data);

    //Get session cookie
    return res.data.sessionCookie as string;
  }

  /**
   * Verify if the provided session cookie is valid.
   * @param sessionCookie JWT session cookie generated from createSessionCookie
   * @returns The decoded JWT payload
   */
  async verifySessionCookie(sessionCookie: string): Promise<DecodedIdToken> {
    //Fetch google public key
    const res = await axios(
      'https://www.googleapis.com/identitytoolkit/v3/relyingparty/publicKeys'
    );

    const header = decodeProtectedHeader(sessionCookie);
    const data = res.data;
    if (!data[header.kid]) throw Error('Cannot find public key');

    //Get certificate from JWT key id
    const certificate = data[header.kid];
    const publicKey = await importX509(certificate, 'RS256');

    //Verify the sessionCookie with the publicKey
    const { payload } = await jwtVerify(sessionCookie, publicKey, {
      issuer: `https://session.firebase.google.com/${this.config.projectId}`,
      audience: this.config.projectId,
    });

    return payload as any as DecodedIdToken;
  }

  /**
   * Verifies a Firebase ID token (JWT).
   * If the token is valid, the promise is fulfilled with the token's decoded claims; otherwise, the promise is rejected.
   * @param idToken An Identity Platform ID token
   */
  async verifyIdToken(idToken: string): Promise<DecodedIdToken> {
    return (await verifyIdToken(idToken)) as any as DecodedIdToken;
  }
}


src/lib/firebase-auth.ts
import { decodeProtectedHeader, importX509, jwtVerify } from 'jose';
import axios from 'redaxios';

import { Cache } from './cache/cache';
import { getAuthToken, verifyIdToken } from './google-oauth';
import { DecodedIdToken, User } from './models';

export type FirebaseConfig = {
  readonly projectId: string;
  readonly apiKey: string;
  readonly serviceAccountEmail: string;
  readonly privateKey: string;
  readonly cache?: Cache;
};

/**
 * Interact with Firebase REST Api and Google Identity Toolkit Api.
 * Made to work with CloudFire Workers
 */
export class FirebaseAuth {
  private BASE_URL = 'https://identitytoolkit.googleapis.com/v1/';

  constructor(public readonly config: FirebaseConfig) {}

  /**
   * Cache the result of an async function
   * @param action Function with result to be stored
   * @param key Where to find/store the value from/to the cache
   * @param expiration Cache expiration in seconds
   * @returns Cached result
   */
  private async withCache<T>(
    action: () => Promise<T>,
    key: string,
    expiration: number
  ): Promise<T> {
    if (!this.config.cache) return await action();

    let result = (await this.config.cache.get(key)) as T;
    if (!result) {
      result = await action();
      await this.config.cache.put(key, result, { expirationTtl: expiration });
    }

    return result;
  }

  /**
   * Cache Result Of Token Response
   * @returns Returns Cached result
   */

  async getToken(): Promise<string> {
    return await this.withCache(
      () =>
        getAuthToken(
          this.config.serviceAccountEmail,
          this.config.privateKey,
          'https://www.googleapis.com/auth/identitytoolkit'
        ),
      'google-oauth',
      3600
    );
  }

  /**
   * Send a post request to the identity toolkit api
   * @param formData POST form data
   * @param endpoint endpoint of the identity toolkit googleapis
   * @returns HTTP Response
   */
  private async sendFirebaseAuthPostRequest(
    formData: Record<string, string>,
    endpoint: string
  ): Promise<Record<string, unknown>> {
    const URI =
      this.BASE_URL + `accounts:${endpoint}?key=${this.config.apiKey}`;

    const res = await axios.post(URI, formData);
    if (res.status != 200) throw Error(res.data);
    return res.data;
  }

  /**
   * Retrieve user info from a Firebase ID token
   * @param idToken A valid Firebase ID token
   * @returns User info linked to this ID token
   */
  public async lookupUser(idToken: string): Promise<User> {
    const data = await this.sendFirebaseAuthPostRequest(
      { idToken: idToken },
      'lookup'
    );

    return data.users[0] as User;
  }

  /**
   * Sign in Firebase user with email and password
   * @param email Email of the Firebase user
   * @param password Password of the Firebase user
   * @returns The decoded JWT token payload and the signed in user info
   */
  async signInWithEmailAndPassword(
    email: string,
    password: string
  ): Promise<{ token: DecodedIdToken; user: User }> {
    const data = await this.sendFirebaseAuthPostRequest(
      {
        email: email,
        password: password,
        returnSecureToken: 'true',
      },
      'signInWithPassword'
    );

    const token = data as DecodedIdToken;
    const user = await this.lookupUser(token.idToken);

    return { token, user };
  }

  /**
   * Change a user's password
   * @param idToken	A Firebase Auth ID token for the user.
   * @param newPassword	User's new password.
   * @returns The decoded JWT token payload
   */
  async changePassword(
    idToken: string,
    newPassword: string
  ): Promise<DecodedIdToken> {
    const data = await this.sendFirebaseAuthPostRequest(
      {
        idToken: idToken,
        password: newPassword,
        returnSecureToken: 'true',
      },
      'update'
    );

    const token = data as DecodedIdToken;

    return token;
  }

  /**
   * Delete a current user
   * @param idToken	A Firebase Auth ID token for the user.
   */
  async deleteAccount(idToken: string) {
    await this.sendFirebaseAuthPostRequest(
      {
        idToken: idToken,
      },
      'delete'
    );
  }

  /**
   * Sign up Firebase user with email and password
   * @param email Email of the Firebase user
   * @param password Password of the Firebase user
   * @returns The decoded JWT token payload and the signed in user info
   */
  async signUpWithEmailAndPassword(
    email: string,
    password: string
  ): Promise<{ token: DecodedIdToken; user: User }> {
    const data = await this.sendFirebaseAuthPostRequest(
      {
        email: email,
        password: password,
        returnSecureToken: 'true',
      },
      'signUp'
    );

    const token = data as DecodedIdToken;
    const user = await this.lookupUser(token.idToken);

    return { token, user };
  }

  /**
   * Sets additional developer claims on an existing user identified by the
   * provided `uid`, typically used to define user roles and levels of
   * access. These claims should propagate to all devices where the user is
   * already signed in (after token expiration or when token refresh is forced)
   * and the next time the user signs in.
   *
   * See {@link https://firebase.google.com/docs/auth/admin/custom-claims |
   * Defining user roles and access levels}
   * for code samples and detailed documentation.
   *
   * @param uid - The `uid` of the user to edit.
   * @param customUserClaims - The developer claims to set. If null is
   *   passed, existing custom claims are deleted. Passing a custom claims payload
   *   larger than 1000 bytes will throw an error. Custom claims are added to the
   *   user's ID token which is transmitted on every authenticated request.
   *   For profile non-access related user attributes, use database or other
   *   separate storage systems.
   * @returns A promise that resolves when the operation completes
   *   successfully.
   */
  public async setCustomUserClaims(
    uid: string,
    customUserClaims: Record<string, unknown> | null
  ): Promise<Record<string, unknown>> {
    if (customUserClaims === null) {
      customUserClaims = {};
    }

    const token = await this.getToken();

    const payload = {
      localId: uid,
      customAttributes: JSON.stringify(customUserClaims),
    };

    const headers = { Authorization: 'Bearer ' + token };

    const path = `projects/${this.config.projectId}/accounts:update`;

    const res = await axios.post(this.BASE_URL + path, payload, { headers });
    if (res.status != 200) throw Error(res.data);

    return res.data;
  }

  /**
   * Creates a session cookie for the given Identity Platform ID token.
   * The session cookie is used by the client to preserve the user's login state.
   * @param idToken A valid Identity Platform ID token
   * @param expiresIn The number of seconds until the session cookie expires.
   * Specify a duration in seconds, between five minutes and fourteen days, inclusively.
   * @returns The session cookie that has been created
   */
  async createSessionCookie(
    idToken: string,
    expiresIn: number = 60 * 60 * 24 * 14 //14 days
  ): Promise<string> {
    //Create the OAuth 2.0 token
    //OAuth token is cached until expiration (1h)
    const token = await this.getToken();

    //Post params and header authorization

    const payload = { idToken: idToken, validDuration: expiresIn + '' };
    const headers = { Authorization: 'Bearer ' + token };

    const path = `projects/${this.config.projectId}:createSessionCookie`;
    const res = await axios.post(this.BASE_URL + path, payload, { headers });
    if (res.status != 200) throw Error(res.data);

    //Get session cookie
    return res.data.sessionCookie as string;
  }

  /**
   * Verify if the provided session cookie is valid.
   * @param sessionCookie JWT session cookie generated from createSessionCookie
   * @returns The decoded JWT payload
   */
  async verifySessionCookie(sessionCookie: string): Promise<DecodedIdToken> {
    //Fetch google public key
    const res = await axios(
      'https://www.googleapis.com/identitytoolkit/v3/relyingparty/publicKeys'
    );

    const header = decodeProtectedHeader(sessionCookie);
    const data = res.data;
    if (!data[header.kid]) throw Error('Cannot find public key');

    //Get certificate from JWT key id
    const certificate = data[header.kid];
    const publicKey = await importX509(certificate, 'RS256');

    //Verify the sessionCookie with the publicKey
    const { payload } = await jwtVerify(sessionCookie, publicKey, {
      issuer: `https://session.firebase.google.com/${this.config.projectId}`,
      audience: this.config.projectId,
    });

    return payload as any as DecodedIdToken;
  }

  /**
   * Verifies a Firebase ID token (JWT).
   * If the token is valid, the promise is fulfilled with the token's decoded claims; otherwise, the promise is rejected.
   * @param idToken An Identity Platform ID token
   * @param customData Public Key Data
   */
  async verifyIdToken(
    idToken: string,
    customData: Record<string, unknown> = {}
  ): Promise<DecodedIdToken> {
    return (await verifyIdToken(idToken, customData)) as any as DecodedIdToken;
  }
}
